web/src/realm/RealmApolloProvider.tsx

```diff
@@ -1,38 +1,57 @@
 import * as React from "react";
-import * as RealmWeb from "realm-web";
+import * as Realm from "realm-web";
 import { useRealmApp } from "./RealmApp";
 
 // Apollo
-import { ApolloClient, HttpLink, InMemoryCache } from "apollo-boost";
-import { setContext } from "apollo-link-context";
-import { ApolloProvider } from "@apollo/react-hooks";
+import {
+  ApolloProvider,
+  ApolloClient,
+  HttpLink,
+  InMemoryCache,
+} from "@apollo/client";
+import { NormalizedCacheObject } from "@apollo/client/cache";
 
 const RealmApolloProvider: React.FC = ({ children }) => {
   const { id, user } = useRealmApp();
   const [client, setClient] = React.useState(
-    createApolloClient(id, user as RealmWeb.User)
+    createApolloClient(id, user as Realm.User)
   );
   React.useEffect(() => {
-    setClient(createApolloClient(id, user as RealmWeb.User));
+    setClient(createApolloClient(id, user as Realm.User));
   }, [id, user]);
 
   return <ApolloProvider client={client}>{children}</ApolloProvider>;
 };
 export default RealmApolloProvider;
 
 // TODO: Implement createApolloClient()
-function createApolloClient(realmAppId: string, user: RealmWeb.User) {
+function createApolloClient(
+  realmAppId: string,
+  user: Realm.User
+): ApolloClient<NormalizedCacheObject> {
   const graphql_url = `https://realm.mongodb.com/api/client/v2.0/app/${realmAppId}/graphql`;
-  const httpLink = new HttpLink({ uri: graphql_url });
-  const authorizationHeaderLink = setContext(async (_, { headers }) => ({
-    headers: {
-      ...headers,
-      Authorization: `Bearer ${user.accessToken}`,
-    },
-  }));
-
-  return new ApolloClient({
-    link: authorizationHeaderLink.concat(httpLink),
+  
+  const client = new ApolloClient({
+    link: new HttpLink({
+      uri: graphql_url,
+      fetch: async (uri: RequestInfo, options: RequestInit) => {
+        if (!options.headers) {
+          options.headers = {} as Record<string, string>;
+        }
+        // Refreshing custom data also ensures a valid access token
+        await user.refreshCustomData();
+        const authenticatedOptions: RequestInit = {
+          ...options,
+          headers: {
+            ...options.headers,
+            Authorization: `Bearer ${user.accessToken}`
```

Comment:
Nitpick: I was curious as to whether ApolloClient had a helper for bearer token headers, so I checked their documentation. As you probably already know, they don't, but they do use this code to protect the token from evaluating as `undefined` which I think might be more appropriate:
```
authorization: token ? `Bearer ${token}` : "",
```

Reply:
Yeah I generally agree that the ternary operator is better here, but there's some TypeScript-specific wonkiness that made this approach much easier (it also made me add the explicit check for `options.headers`)